src/main/java/com/gojek/beast/sink/bq/handler/gcs/GCSErrorWriter.java

```diff
@@ -0,0 +1,99 @@
+package com.gojek.beast.sink.bq.handler.gcs;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.gojek.beast.config.Constants;
+import com.gojek.beast.sink.bq.handler.impl.BQErrorHandlerException;
+import com.gojek.beast.sink.bq.handler.ErrorWriter;
+import com.gojek.beast.models.OffsetInfo;
+import com.gojek.beast.models.Record;
+import com.gojek.beast.models.Status;
+import com.gojek.beast.sink.bq.handler.WriteStatus;
+import com.gojek.beast.stats.Stats;
+import com.gojek.beast.util.ApplicationUtil;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Blob;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@AllArgsConstructor
+@Slf4j
+public class GCSErrorWriter implements ErrorWriter {
+
+    private final Storage gcsStore;
+    private final String gcsBucket; // <bucket> to store
+    private final String gcsBasePathPrefix; // path prefix
+    private final Stats statsClient = Stats.client(); // metrics client
+    private final ObjectMapper mapper = new ObjectMapper();
+
+    @Override
+    public Status writeErrorRecords(List<Record> errorRecords) {
+        if (errorRecords.isEmpty()) {
+            return new WriteStatus(true, Optional.ofNullable(null));
+        }
+        final Instant startTime = Instant.now();
+        try {
+            //Store messages in the GCS
+            storeMessagesInGCS(errorRecords);
+        } catch (StorageException se) {
+            log.error("Exception::Failed to write to GCS: {} records size: {}", se, errorRecords.size());
+            throw new BQErrorHandlerException(se.getMessage());
+        }
+        statsClient.gauge("sink.gcs.push.invalid.records", errorRecords.size());
+        statsClient.timeIt("sink.gcs.push.invalid.time", startTime);
+        //alter the insert status - as successful
+        return new WriteStatus(true, Optional.ofNullable(null));
+    }
+
+    private void storeMessagesInGCS(final List<Record> errorRecords) throws StorageException {
```

Comment:
I'm also curious, why do we need to write to GCS everytime, as it'll be incur latency, if it's similar to deadletter, why can't we aggregate it inmemory and write it in batches in separate thread (also during crash)

Reply:
Valid point. We discussed this approach in the team before the implementation, and we decided to proceed with synchronous handling as we have to handle the offset commits in the current batch with all the permutations. Also the invalid messages as I hear occurs rare and essentially are low in numbers, so essentially a negligible from a performance point of view.